_posts/append-only.md

```diff
@@ -0,0 +1,104 @@
+---
+layout: post
+title: "Append-only models: The why, the when, and the how"
+author: Mauro Servienti
+synopsis: "There are scenarios where updating data is either forbidden by policies or undesirable because it hinders the ability to fulfill other requirements. That is when append-only models come to the rescue."
+header_image: /img/posts/append-only/header.jpg
+tags:
+- architecture
+---
+
+In ["All our aggregates are wrong,"](https://youtu.be/KkzvQSuYd5I), I present a distributed system architecture for a sample shopping cart.
+
+The primary goal of the talk is to discuss [coupling and its implications](https://milestone.topics.it/2023/05/17/back-to-basics-boundaries.html), especially in distributed systems.
+
+The secondary one is to present attendees with a more nuanced distributed architecture design and introduce [ViewModel Composition](https://milestone.topics.it/series/view-model-composition.html) as a solution to some of the concerns a distributed architecture poses.
+
+During the talk, I also mentioned [ViewModel Decomposition](https://milestone.topics.it/2019/04/18/the-fine-art-of-dismantling.html) and why we might need it. In that context, I refer to append/insert-only models as the way to go when dealing with those scenarios.
+
+Recently, [Domenic](https://www.linkedin.com/in/domenic-cassisi/) shared the observation that append-only models are a rarity, and he sees updates used everywhere every day. And he's right!
+
+Append-only models are more complex to manage. In fact, in most cases, projections are a must-have to satisfy read requirements. That's to say, I'm not surprised there isn't widespread adoption. And that's just fine; it doesn't need to be.
+
+## Mauro, what are you talking about?
+
+Let's start by trying to define what append-only means. Generally speaking, an append-only model or approach is a scenario in which persisting data is never achieved using `update` statements but only `insert` ones.
+
+A bank statement is something we have seen at least once. Banks never update statements; instead, they add data to represent money transactions or compensations.
+
+Why would we need such a data-storing mechanism besides banks or accounting?
+
+Let's imagine a warehouse management system. To store data, we could use a schema like the following:
+
+| PK | Description | Quantity | Purchase price |
+|----|-------------|----------|----------------|
+|abc |  Something  |       32 |         123.00 |
+
+If you have ever dealt with an accounting system, you probably already see the issue. One of the requirements is to run an inventory using a method like LIFO to calculate warehouse value. Ah, crap! We don't have any of the needed information.
```

Comment:
Consider linking LIFO to an article on what it means in case it's not well-known. And you might to explain "We don't have any of the needed information". E.g. 

```suggestion
If you have ever dealt with an accounting system, you probably already see the issue. One of the requirements is to run an inventory using a method like LIFO to calculate warehouse value. I.e. we want the most recently purchased items that were added to our inventory to be the first ones out when an order comes in. Ah, crap! We don't have any of the needed information.
```

Reply:
```suggestion
If you have ever dealt with an accounting system, you probably already see the issue. One of the requirements is to run an inventory using a method like LIFO ([last-in-first-out](https://www.forbes.com/advisor/business/lifo-inventory-method/)) to calculate warehouse value. I.e., we want the most recently purchased items that were added to our inventory to be the first ones out when an order comes in. Ah, crap! We don't have any of the needed information.
```